### ISSUE TITLE
A Simple Explanation of Gini Impurity - victorzhou.com

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8632]
A Simple Explanation of Gini Impurity - victorzhou.com<br>
<br>
https://ift.tt/sMB3tAz<br>
<br>
Victor Zhou<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>If you look at the documentation for the <a rel="noopener noreferrer" href="https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html" target="_blank">DecisionTreeClassifier</a> class in <a rel="noopener noreferrer" href="https://scikit-learn.org" target="_blank">scikit-learn</a>, you&rsquo;ll see something like this for the <code>criterion</code> parameter:</p>
<p><span><img alt="scikit learn" title="scikit learn" src="https://victorzhou.com/static/49768553077e5fd43f35c774534779d8/39600/scikit-learn.png"></span></p>
<p>The <a rel="noopener noreferrer" href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html" target="_blank">RandomForestClassifier</a> documentation says the same thing. Both mention that the default criterion is &ldquo;gini&rdquo; for the <strong>Gini Impurity</strong>. What is that?!</p>
<blockquote>
<p>TLDR: Read the <a href="https://victorzhou.com/blog/gini-impurity/#recap">Recap</a>.</p>
</blockquote>
<h2>
<a href="https://victorzhou.com/blog/gini-impurity/#decision-trees-"></a>Decision Trees &#127794;</h2>
<p>Training a decision tree consists of iteratively splitting the current data into two branches. Say we had the following datapoints:</p>
<figure><img src="https://victorzhou.com/0a81728a87e7cf55cf58a653f9c210c5/dataset.svg">
<figcaption>The Dataset</figcaption>
</figure>
<p>Right now, we have 1 branch with 5 blues and 5 greens.</p>
<p>Let&rsquo;s make a split at :</p>
<figure><img src="https://victorzhou.com/d99afd724f34af430a26dd0670b57cf3/dataset-perfect-split.svg">
<figcaption>A Perfect Split</figcaption>
</figure>
<p>This is a <strong>perfect</strong> split! It breaks our dataset perfectly into two branches:</p>
<ul>
<li>Left branch, with 5 blues.</li>
<li>Right branch, with 5 greens.</li>
</ul>
<p>What if we&rsquo;d made a split at instead?</p>
<figure><img src="https://victorzhou.com/4768c79349a62634d54ce531b3680751/dataset-imperfect-split.svg">
<figcaption>An Imperfect Split</figcaption>
</figure>
<p>This imperfect split breaks our dataset into these branches:</p>
<ul>
<li>Left branch, with 4 blues.</li>
<li>Right branch, with 1 blue and 5 greens.</li>
</ul>
<p>It&rsquo;s obvious that this split is worse, but <strong>how can we quantify that?</strong></p>
<p>Being able to measure the quality of a split becomes even more important if we add a third class, reds . Imagine the following split:</p>
<ul>
<li>Branch 1, with 3 blues, 1 green, and 1 red.</li>
<li>Branch 2, with 3 greens and 1 red.</li>
</ul>
<p>Compare that against this split:</p>
<ul>
<li>Branch 1, with 3 blues, 1 green, and 2 reds.</li>
<li>Branch 2, with 3 greens.</li>
</ul>
<p>Which split is better? It&rsquo;s no longer immediately obvious. We need a way to <strong>quantitatively evaluate</strong> how good a split is.</p>
<h2>
<a href="https://victorzhou.com/blog/gini-impurity/#gini-impurity"></a>Gini Impurity</h2>
<p>This is where the Gini Impurity metric comes in.</p>
<p>Suppose we</p>
<ol>
<li>Randomly pick a datapoint in our dataset, then</li>
<li>
<strong>Randomly classify it according to the class distribution in the dataset</strong>. For our dataset, we&rsquo;d classify it as blue of the time and as green of the time, since we have 5 datapoints of each color.</li>
</ol>
<p><strong>What&rsquo;s the probability we classify the datapoint incorrectly?</strong> The answer to that question is the Gini Impurity.</p>
<h3>
<a href="https://victorzhou.com/blog/gini-impurity/#example-1-the-whole-dataset"></a>Example 1: The Whole Dataset</h3>
<p>Let&rsquo;s calculate the Gini Impurity of our entire dataset. If we randomly pick a datapoint, it&rsquo;s either blue (50%) or green (50%).</p>
<p>Now, we randomly classify our datapoint according to the class distribution. Since we have 5 of each color, we classify it as blue 50% of the time and as green 50% of the time.</p>
<p>What&rsquo;s the probability we classify our datapoint <strong>incorrectly</strong>?</p>
<table>
<thead>
<tr>
<th>Event</th>
<th>Probability</th>
</tr>
</thead>
<tbody>
<tr>
<td>Pick Blue, Classify Blue <span>&#10003;</span>
</td>
<td>25%</td>
</tr>
<tr>
<td>Pick Blue, Classify Green &#10060;</td>
<td>25%</td>
</tr>
<tr>
<td>Pick Green, Classify Blue &#10060;</td>
<td>25%</td>
</tr>
<tr>
<td>Pick Green, Classify Green <span>&#10003;</span>
</td>
<td>25%</td>
</tr>
</tbody>
</table>
<p>We only classify it incorrectly in 2 of the events above. Thus, our total probability is 25% + 25% = 50%, so the Gini Impurity is .</p>
<h3>
<a href="https://victorzhou.com/blog/gini-impurity/#the-formula"></a>The Formula</h3>
<p>If we have total classes and is the probability of picking a datapoint with class , then the Gini Impurity is calculated as</p>
<p>For the example above, we have and , so</p>
<p>which matches what we calculated!</p>
<h3>
<a href="https://victorzhou.com/blog/gini-impurity/#example-2-a-perfect-split"></a>Example 2: A Perfect Split</h3>
<p>Let&rsquo;s go back to the perfect split we had. What are the Gini Impurities of the two branches after the split?</p>
<figure><img src="https://victorzhou.com/d99afd724f34af430a26dd0670b57cf3/dataset-perfect-split.svg">
<figcaption>A Perfect Split</figcaption>
</figure>
<p>Left Branch has only blues, so its Gini Impurity is</p>
<p>Right Branch has only greens, so its Gini Impurity is</p>
<p>Both branches have impurity! The perfect split turned a dataset with impurity into 2 branches with impurity.</p>
<p><strong>A Gini Impurity of 0 is the lowest and best possible impurity</strong>. It can only be achieved when everything is the same class (e.g. only blues or only greens).</p>
<h3>
<a href="https://victorzhou.com/blog/gini-impurity/#example-3-an-imperfect-split"></a>Example 3: An Imperfect Split</h3>
<p>Finally, let&rsquo;s return to our imperfect split.</p>
<figure><img src="https://victorzhou.com/4768c79349a62634d54ce531b3680751/dataset-imperfect-split.svg">
<figcaption>An Imperfect Split</figcaption>
</figure>
<p>Left Branch has only blues, so we know that .</p>
<p>Right Branch has 1 blue and 5 greens, so</p>
<h2>
<a href="https://victorzhou.com/blog/gini-impurity/#picking-the-best-split"></a>Picking The Best Split</h2>
<p>It&rsquo;s finally time to answer the question we posed earlier: <strong>how can we quantitatively evaluate the quality of a split?</strong></p>
<p>Here&rsquo;s the imperfect split yet again:</p>
<figure><img src="https://victorzhou.com/4768c79349a62634d54ce531b3680751/dataset-imperfect-split.svg">
<figcaption>An Imperfect Split</figcaption>
</figure>
<p>We&rsquo;ve already calculated the Gini Impurities for:</p>
<ul>
<li>Before the split (the entire dataset):</li>
<li>Left Branch:</li>
<li>Right Branch:</li>
</ul>
<p>We&rsquo;ll determine the quality of the split by <strong>weighting the impurity of each branch by how many elements it has</strong>. Since Left Branch has 4 elements and Right Branch has 6, we get:</p>
<p>Thus, the amount of impurity we&rsquo;ve &ldquo;removed&rdquo; with this split is</p>
<p>I&rsquo;ll call this value the Gini Gain. This is what&rsquo;s used to pick the best split in a decision tree! <strong>Higher Gini Gain = Better Split</strong>. For example, it&rsquo;s easy to verify that the Gini Gain of the perfect split on our dataset is .</p>
<h2>
<a href="https://victorzhou.com/blog/gini-impurity/#recap"></a>Recap</h2>
<p><strong>Gini Impurity</strong> is the probability of <em>incorrectly</em> classifying a randomly chosen element in the dataset if it were randomly labeled <em>according to the class distribution</em> in the dataset. It&rsquo;s calculated as</p>
<p>where is the number of classes and is the probability of randomly picking an element of class .</p>
<p>When training a decision tree, the best split is chosen by <strong>maximizing the Gini Gain</strong>, which is calculated by subtracting the weighted impurities of the branches from the original impurity.</p>
<p>Want to learn more? Check out my explanation of <a href="https://victorzhou.com/blog/information-gain/">Information Gain</a>, a similar metric to Gini Gain, or my guide <a href="https://victorzhou.com/blog/intro-to-random-forests/">Random Forests for Complete Beginners</a>.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via victorzhou.com<br>
<br>
August 26, 2022 at 06:22PM
